### PR TITLE
feat: raycaster picker

### LINKED ISSUE #59: Mac users cannot click on nodes
Clicking (both left and right clicks) on nodes does not work on certain Macs.

This used to be an obscure issue that only affected a specific MacBook but recently shown up in two newer MacBooks (both with the M1 chip). 

Message about potential work around: https://github.com/MetabolicAtlas/MetabolicAtlas/issues/1144#issuecomment-1290165202

### PR DESCRIPTION
**Pull request by e0**
**Related issue(s) and PR(s)**  
This PR closes #59.

**Type of change**  

- [x] Bug fix (non-breaking change which fixes an issue)
 
**List of changes made**  

Use [`Raycaster#intersectObjects`](https://threejs.org/docs/#api/en/core/Raycaster.intersectObject) instead of [`WebGLRenderer#readRenderTargetPixels`](https://threejs.org/docs/#api/en/renderers/WebGLRenderer.readRenderTargetPixels) to pick nodes.

**Screenshot of the fix**  

image

**Testing**  

- A certain branch in other repos is needed
- A rebuild is needed due to new dependencies
- Relative urls that can be reused both for production and local testing
- Instructions on how to test

For real world use, it is best to test this branch using the [MetabolicAtlas repo](https://github.com/MetabolicAtlas/MetabolicAtlas).

- To use it, do:
- 1. Change value for dependency from version number to git branch in `frontend/package.json`.
```json
    "@metabolicatlas/3d-network-viewer": "MetabolicAtlas/3d-network-viewer#feat/raycaster-picker",
```

- 2.  Install the dependency and build it.
```bash
docker-compose --env-file env-local.env -f docker-compose.yml -f docker-compose-local.yml up -d --no-deps --build frontend
ma-exec frontend npm --prefix node_modules/@metabolicatlas/3d-network-viewer/ install
ma-exec frontend npm --prefix node_modules/@metabolicatlas/3d-network-viewer/ run build
```
- Try out the 3d maps and the interaction partners, apply overlay, select nodes, search for nodes, reload the page etc.

**Further comments**  

This seems to work fine for most cases. I have noticed that for nodes that are very close together and on top of each other, sometimes the wrong node can be selected. This is certainly a drawback compared to the previous implementation. I have spent quite a few hours trying to improve this but it's still problematic sometimes.

Since it's unclear if spending more time would improve this right now, I suggest that we move forward with one of the following two options. It would be great to get more testing and feedback on this as well.

1. Implement this.
  - **pro**: Make node picking work for users with Apple silicon chips as well as some other GPUs. 
  - **con**: Worse experience for other users.

2. Do not implement this.
  - **pro**: Do not degrade performance for other users.
  - **con**: Leave users with Apple silicon chips (and some others) hanging.

**Definition of Done checklist**  
- [x] My code meets the Acceptance Criteria
- [x] My code follows the [NBIS style guidelines](https://github.com/NBISweden/development-guidelines)
- [x] I have performed a self-review of my own code and commented any hard-to-understand areas
- [x] Tests and lint/format validations are passing
- [x] My changes generate no new warnings

### CODE REVIEW

File: src/raycaster-picker.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+  pick({ renderer, scene, camera, event }) {
+    const { raycaster, pointer } = this;
+    const { clientX, clientY } = event;
+    const rect = renderer.domElement.getBoundingClientRect();
+
+    pointer.x = ((clientX - rect.left) / (rect.right - rect.left)) * 2 - 1;
+    pointer.y = -((clientY - rect.top) / (rect.bottom - rect.top)) * 2 + 1;
+
+    raycaster.setFromCamera(pointer, camera);
+    const intersects = raycaster.intersectObjects(scene.children, true);
```
Comment (mihai-sysbio) [inline src/raycaster-picker.js:19]: Random question: is it just the nodes that are `scene.children`, or is it edges too - can edge selection be interfering in the node selection process?

**Comment by nanjiangshu** [comment]
Great implementation @e0, I tested the 3d-network viewer with the example network and it seems the node selection works very well in most cases, although I did find it weird in some cases as shown in the screen recording below, which shows a square node is selected when clicking a circular node, and also not able to select another circular node. 

https://user-images.githubusercontent.com/5833722/202928559-27f346cc-a29e-48a9-b43c-b45db5c0c2f8.mov

I tried also to use this fix in the 3d-network-viewer in Map Viewer and Interaction partners. However, after installing this branch by 
changing the `frontend/package.json` to `"@metabolicatlas/3d-network-viewer": "MetabolicAtlas/3d-network-viewer#feat/raycaster-picker", ` and then run 
```
ma-exec frontend yarn
ma-exec frontend npm --prefix node_modules/@metabolicatlas/3d-network-viewer/ install
ma-exec frontend npm --prefix node_modules/@metabolicatlas/3d-network-viewer/ run build 
```
it still does not work for the 3D Map viewer, nor for the Interaction Partners. 
It seems the new code has been installed since I got 

```
$ docker exec -it frontend cat node_modules/@metabolicatlas/3d-network-viewer/src/met-atlas-viewer.js | grep -i raycasterPicker
import { RaycasterPicker } from './raycaster-picker';
  let raycasterPicker;
    raycasterPicker = new RaycasterPicker(nodeSize);
    raycasterPicker.pick({
```

**Comment by e0** [comment]
Thanks for testing and including the video @nanjiangshu! The problem from the example seems pretty bad imo. I didn't encounter anything that selected from that far away from the cursor when I tested so it's really nice that we got this example.

For the installation, please replace `ma-exec frontend yarn` with `docker-compose --env-file env-local.env -f docker-compose.yml -f docker-compose-local.yml up -d --no-deps --build frontend
`.

**Comment by nanjiangshu** [comment]
It works well for the node selection now after rebuilding with `docker-compose --env-file env-local.env -f docker-compose.yml -f docker-compose-local.yml up -d --no-deps --build frontend`.
I also found the problem that @MalinAhlberg  mentioned this morning, i.e. the color of nodes turning white. Could someone please check this bug is introduced by https://github.com/MetabolicAtlas/3d-network-viewer/pull/63 or earlier PRs?

**Comment by e0** [comment]
Thanks for testing again @nanjiangshu!

I think checking this other bug could be done in [the new issue](https://github.com/MetabolicAtlas/MetabolicAtlas/issues/1221) since it's unrelated to this change.

**Comment by MalinAhlberg** [comment]
I also tested this, and it works very well for me, well done :+1: . No weird behaviors spotted. It would be interesting to know more about how often the problems occur, though, before deciding whether to implement this or not. From my own testing, I would say it definitely worth having, but it depends on how often others encounter problems.

**Comment by nanjiangshu** [comment]
I think the problem I showed in the screen recording I attached is not serious. For almost all IP networks and most of the networks in 3D Map Viewer, node selection works as expected.

**Review by nanjiangshu — APPROVED** [review]

**Review by MalinAhlberg — APPROVED** [review]
Thanks for the info, @e0 ! I tested clicking around on cytosol, but noticed nothing weird there either, so I approve. Good job :star:!

**Pull request merged**